File: Platforms/Common/Reign_Video_D3D11_Component/Buffers/BufferLayout.cpp
#include "pch.h"
#include "BufferLayout.h"

namespace Reign_Video_D3D11_Component
{
	#pragma region Constructors
	BufferLayoutErrors BufferLayoutCom::Init(VideoCom^ video, ShaderModelCom^ vertexShader, BufferLayoutDescCom^ desc)
	{
		null();
		this->video = video;

		ID3D11InputLayout* layoutTEMP;
		#if WINDOWS
		if (FAILED(video->device->CreateInputLayout(desc->desc, desc->elementCount, vertexShader->code->GetBufferPointer(), vertexShader->code->GetBufferSize(), &layoutTEMP)))
		#else
		if (FAILED(video->device->CreateInputLayout(desc->desc, desc->elementCount, vertexShader->code, vertexShader->codeSize, &layoutTEMP)))
		#endif
		{
			return BufferLayoutErrors::InputLayout;
		}
		layout = layoutTEMP;

		return BufferLayoutErrors::None;
	}

	BufferLayoutCom::~BufferLayoutCom()
	{
		if (layout) layout->Release();
		null();
	}

	void BufferLayoutCom::null()
	{
		layout = 0;
	}
	#pragma endregion
	
	#pragma region Methods
	void BufferLayoutCom::Enable()
	{
		video->deviceContext->IASetInputLayout(layout);
	}
	#pragma endregion
}

File: Platforms/Common/Reign.Video.D3D11/Buffers/BufferLayout.cs
﻿using Reign.Core;
using Reign_Video_D3D11_Component;
using System;

namespace Reign.Video.D3D11
{
	public class BufferLayout : DisposableResource, IBufferLayout
	{
		#region Properties
		private BufferLayoutCom com;
		#endregion

		#region Constructors
		public BufferLayout(IDisposableResource parent, IShader shader, IBufferLayoutDesc desc)
		: base(parent)
		{
			try
			{
				var video = parent.FindParentOrSelfWithException<Video>();
				com = new BufferLayoutCom();
				var error = com.Init(video.com, ((Shader)shader).vertex.com, ((BufferLayoutDesc)desc).com);

				if (error == BufferLayoutErrors.InputLayout) Debug.ThrowError("BufferLayout", "Failed to create InputLayout");
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (com != null)
			{
				com.Dispose();
				com = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public void Enable()
		{
			com.Enable();
		}
		#endregion
	}
}